src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs

```diff
@@ -0,0 +1,122 @@
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    internal class CSharpUpgradeProjectCodeFixProvider : AbstractUpgradeProjectCodeFixProvider
+    {
+        private const string CS8022 = nameof(CS8022); // error CS8022: Feature is not available in C# 1.  Please use language version X or greater.
+        private const string CS8023 = nameof(CS8023); // error CS8023: Feature is not available in C# 2.  Please use language version X or greater.
+        private const string CS8024 = nameof(CS8024); // error CS8024: Feature is not available in C# 3.  Please use language version X or greater.
+        private const string CS8025 = nameof(CS8025); // error CS8025: Feature is not available in C# 4.  Please use language version X or greater.
+        private const string CS8026 = nameof(CS8026); // error CS8026: Feature is not available in C# 5.  Please use language version X or greater.
+        private const string CS8059 = nameof(CS8059); // error CS8059: Feature is not available in C# 6.  Please use language version X or greater.
+
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(CS8022, CS8023, CS8024, CS8025, CS8026, CS8059);
+
+        protected override ImmutableArray<CodeAction> GetUpgradeProjectCodeActionsAsync(CodeFixContext context)
+        {
+            var project = context.Document.Project;
+            var solution = project.Solution;
+            var newVersion = SuggestedVersion(context.Diagnostics);
+
+            var fixOneProjectTitle = string.Format(CSharpFeaturesResources.Upgrade_project_to_csharp_language_version_0,
+                newVersion.Display());
+
+            var fixOneProject = new SolutionChangeAction(fixOneProjectTitle, ct => UpgradeProject(project, solution, newVersion));
+
+            var fixAllProjectsTitle = string.Format(CSharpFeaturesResources.Upgrade_all_projects_to_csharp_language_version_0,
+                newVersion.Display());
+
+            var fixAllProjects = new SolutionChangeAction(fixAllProjectsTitle, ct => UpgradeProjects(solution, newVersion));
+
+            return new CodeAction[] { fixOneProject, fixAllProjects }.AsImmutable();
+        }
+
+        private LanguageVersion SuggestedVersion(ImmutableArray<Diagnostic> diagnostics)
+        {
+            return diagnostics.Select(d => SuggestedVersion(d.Id)).First();
+        }
+
+        private LanguageVersion SuggestedVersion(string id)
+        {
+            switch (id)
+            {
+                case CS8022:
+                case CS8023:
+                case CS8024:
+                case CS8025:
+                case CS8026:
+                case CS8059:
+                    return LanguageVersion.Default;
+                default:
+                    return LanguageVersion.Latest;
+            }
+        }
+
+        private static Task<Solution> UpgradeProject(Project project, Solution solution, LanguageVersion version)
```

Comment:
This is really confusing that you're passing both the project and solution, especially in the UpgradeProjects case where they are from different snapshots.